File: grab/chrome.go
package grab

import (
	"douyin-grab/pkg/logger"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"time"

	"github.com/tebeka/selenium"
	"github.com/tebeka/selenium/chrome"
	slog "github.com/tebeka/selenium/log"
)

type DYMessage struct {
	Message struct {
		Method string `json:"method"`
		Params struct {
			Url string `json:"url"`
		} `json:"params"`
	} `json:"message"`
}

func GetWssUrl(liveRoomUrl string) (string, error) {
	var (
		chromeDriverPath = os.Getenv("CHROME_DRIVER_PATH")
		chromePath       = os.Getenv("CHROME_PATH")
		port, _          = strconv.Atoi(os.Getenv("CHROME_DRIVER_SERVICE_PORT"))
	)

	options := []selenium.ServiceOption{}
	//selenium.SetDebug(true)
	service, err := selenium.NewChromeDriverService(chromeDriverPath, port, options...)
	if err != nil {
		logger.Error("selenium new chrome driver service error %s", err.Error())
		return "", err
	}
	defer service.Stop()

	caps := selenium.Capabilities{
		"browserName": "chrome",
		"goog:loggingPrefs": map[string]string{ //can get log with this, important!!
			"performance": "ALL",
		},
	}
	chromeCaps := chrome.Capabilities{
		Path: chromePath,
		Args: []string{
			"--headless",
			"--disable-gpu",
			"--no-sandbox",
			"--disable-dev-shm-usage",
			"--disable-background-timer-throttling",
			"--disable-backgrounding-occluded-windows",
			"--disable-renderer-backgrounding",
			"user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3",
		},
	}
	caps.AddChrome(chromeCaps)
	webDriver, err := selenium.NewRemote(caps, fmt.Sprintf("http://localhost:%d/wd/hub", port))
	if err != nil {
		logger.Error("selenium new remote error %s", err.Error())
		return "", err
	}
	defer webDriver.Quit()

	//start request
	_ = webDriver.Get(liveRoomUrl)

	//wait web page loading
	time.Sleep(2 * time.Second)

	//get request log
	logs, err := webDriver.Log(slog.Performance)
	if nil == err {
		if nil == err {
			for _, message := range logs {
				// fmt.Println(message.Message)
				var msg DYMessage
				err = json.Unmarshal([]byte(message.Message), &msg)
				if nil == err {
					if msg.Message.Method == "Network.webSocketCreated" {
						wssurl := msg.Message.Params.Url

						return wssurl, nil
					}
				} else {
					return "", err
				}

				// file, err := os.OpenFile("output.txt", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
				// if err != nil {
				// 	log.Fatal(err)
				// }
				// defer file.Close()
				// _, err = file.WriteString(message.Message)
				// if err != nil {
				// 	log.Fatal(err)
				// }
			}
		}
	}

	// 截屏
	//img, _ := webDriver.Screenshot()
	//_ = ioutil.WriteFile("douyin.png", img, 0666)

	return "", nil
}

File: grab/chrome.py
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities
import time

# 启动Chrome浏览器，并打开WSS调试端口
chrome_options = webdriver.ChromeOptions()
chrome_options.add_argument("headless")             # 无界面
chrome_options.add_argument("--no-sandbox")           # root用户启动需要开启
chrome_options.add_argument("--disable-dev-shm-usage")
chrome_options.add_argument("--remote-debugging-port=9222")
chrome_options.add_argument("--disable-background-timer-throttling")
chrome_options.add_argument("--disable-backgrounding-occluded-windows")
chrome_options.add_argument("--disable-renderer-backgrounding")
chrome_options.add_argument("user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3") #需要agnet,不然一些网站会有限制
#chrome_options.binary_location = "/usr/bin/chromium-browser"  #linux下需要

# service = Service("/path/to/chromedriver") #linux下需要
service = Service("D:\chromedriver.exe") #windows下需要
caps = DesiredCapabilities.CHROME
caps["goog:loggingPrefs"] = {"performance": "ALL"}
driver = webdriver.Chrome(service=service, options=chrome_options, desired_capabilities=caps) #selenium4方式

# 等待页面加载完成
driver.get("https://live.douyin.com/80017709309")
time.sleep(2)

# 获取所有的网络请求日志信息
logs = driver.get_log("performance")

# 打印所有的WSS请求信息
for log in logs:
    message = log["message"]
    if "webSocket" in message:
        print(message)

driver.quit()